demo/flutter/lib/mic_recorder.dart

```diff
@@ -0,0 +1,163 @@
+import 'package:flutter/services.dart';
+import 'package:flutter_voice_processor/flutter_voice_processor.dart';
+import 'package:leopard_flutter/leopard_error.dart';
+import 'package:path_provider/path_provider.dart';
+
+typedef RecordedCallback = Function(double length);
+typedef ProcessErrorCallback = Function(LeopardException error);
+
+class MicRecorder {
+  final VoiceProcessor? _voiceProcessor;
+  int _sampleRate;
+
+  final RecordedCallback _recordedCallback;
+  final ProcessErrorCallback _processErrorCallback;
+  RemoveListener? _removeVoiceProcessorListener;
+  RemoveListener? _removeErrorListener;
+
+  List<int> _pcmData = [];
+
+  static Future<MicRecorder> create(int sampleRate, RecordedCallback recordedCallback, ProcessErrorCallback processErrorCallback) async {
+    return MicRecorder._(sampleRate, recordedCallback, processErrorCallback);
+  }
+
+  MicRecorder._(this._sampleRate, this._recordedCallback, this._processErrorCallback)
+      : _voiceProcessor = VoiceProcessor.getVoiceProcessor(512, _sampleRate) {
+    if (_voiceProcessor == null) {
+      throw LeopardRuntimeException("flutter_voice_processor not available.");
+    }
+    _removeVoiceProcessorListener =
+        _voiceProcessor!.addListener((buffer) async {
+      List<int> frame;
+      try {
+        frame = (buffer as List<dynamic>).cast<int>();
+      } on Error {
+        LeopardException castError = LeopardException(
+            "flutter_voice_processor sent an unexpected data type.");
+        _processErrorCallback(castError);
+        return;
+      }
+
+      _pcmData.addAll(frame);
+      _recordedCallback(_pcmData.length / _sampleRate);
+    });
+
+    _removeErrorListener = _voiceProcessor!.addErrorListener((errorMsg) {
+      LeopardException nativeError = LeopardException(errorMsg as String);
+      _processErrorCallback(nativeError);
+    });
+  }
+
+  Future<void> startRecord() async {
+    if (_voiceProcessor == null) {
+      throw LeopardInvalidStateException(
+          "Cannot start audio recording - resources have already been released");
+    }
+
+    _pcmData.clear();
+
+    if (await _voiceProcessor?.hasRecordAudioPermission() ?? false) {
+      try {
+        await _voiceProcessor!.start();
+      } on PlatformException {
+        throw LeopardRuntimeException(
+            "Audio engine failed to start. Hardware may not be supported.");
+      }
+    } else {
+      throw LeopardRuntimeException(
+          "User did not give permission to record audio.");
+    }
+  }
+
+  Future<File> stopRecord() async {
+    if (_voiceProcessor == null) {
+      throw LeopardInvalidStateException(
+          "Cannot stop audio recording - resources have already been released");
+    }
+
+    if (_voiceProcessor?.isRecording ?? false) {
+      await _voiceProcessor!.stop();
+    }
+
+    try {
+      return await writeWavFile();
+    } catch(e) {
+      throw LeopardIOException(
+          "Failed to save recorded aaudio to file.");
```

Comment:
```suggestion
          "Failed to save recorded audio to file.");
```

Reply:
Ugh, my keyboaard keeps double typing `A`s it's very annoying